### PR TITLE
Fixes #21083  - Link tasks to users explicitly

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

**Comment by adamruzicka** [comment]
This will probably deprecate https://github.com/theforeman/foreman-tasks/pull/363

**Comment by iNecas** [comment]
Test failures related

File: app/models/foreman_tasks/lock.rb [hunk 1]
```diff
@@ -120,7 +120,7 @@ def link?(resource, uuid)
 
       # Records the information about the user that triggered the task
       def owner!(user, uuid)
-        build_owner(user, uuid).save!
```
Comment (iNecas) [inline app/models/foreman_tasks/lock.rb]: Any reason to keep this around https://github.com/theforeman/foreman-tasks/blob/4aa960623bfb19dff930b96efcf23bab7019a90b/lib/foreman_tasks/dynflow/persistence.rb#L39 ?

Comment (adamruzicka) [reply app/models/foreman_tasks/lock.rb]: I can't think of any, just an oversight on my side

**Comment by adamruzicka** [comment]
test failures seem unrelated

File: test/factories/task_factory.rb [hunk 2]
```diff
@@ -6,11 +6,11 @@
     result 'success'
 
     transient do
-      set_owner nil
+      lock_owner nil
     end
 
     after(:create) do |task, evaluator|
-      ForemanTasks::Lock.owner!(evaluator.set_owner, task.id) if evaluator.set_owner
+      ForemanTasks::Lock.owner!(evaluator.lock_owner, task.id) if evaluator.lock_owner
```
Comment (iNecas) [inline test/factories/task_factory.rb]: Is the lock_owner used anywhere? If not, shouldn't we just remove it?

Comment (adamruzicka) [reply test/factories/task_factory.rb]: I briefly grep'd through the code and some plugins and couldn't find it (or the original variant) being uesd anywhere. Dropped

File: app/models/foreman_tasks/lock.rb [hunk 3]
```diff
@@ -120,7 +120,7 @@ def link?(resource, uuid)
...
+        # TODO: Log some deprecation warning?
```
Comment (iNecas) [inline app/models/foreman_tasks/lock.rb]: Can we drop this method all together with other unused code or still the todo is valid?

File: db/migrate/20180927120509_add_user_id.foreman_tasks.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        user_locks.where(:resource_id => owner_id).delete_all
+      end
+    end
+  end
+
+  def down
+    User.as_anonymous_admin do
+      ForemanTasks::Task.select(:user_id).distinct.pluck(:user_id).compact.each do |user_id|
+        ForemanTasks::Task.where(:user_id => user_id).select(:id).find_in_batches do |group|
+          group.each { |task| create_lock user_id, task } # TODO: Bulk insert would be nice
```
Comment (iNecas) [inline db/migrate/20180927120509_add_user_id.foreman_tasks.rb]: Given the probability of the down scenario, I'm ok with removing this TODO, as it's probably not worth doing it.

**Comment by iNecas** [comment]
I have not tested yet, but code-wise looks good. I would like to wait until the branch date of Foreman 1.20: it's still relatively big change and I would be more comfortable having this in nightlies for a while before we go GA with it.

**Comment by adamruzicka** [comment]
Agreed, let's put this on hold for a while

**Comment by adamruzicka** [comment]
[test foreman-tasks]

**Comment by ares** [comment]
:clap:

**Comment by adamruzicka** [comment]
The tests are green, let's put this on hold until all the 1.20 candidates are dealt with

**Comment by ares** [comment]
needs rebase

**Comment by iNecas** [comment]
Release the kraken!

**Pull request merged**